File: Codelabs(V2)/Lesson-10/10.1_RoomWordsSample/app/src/main/java/dev/emg/roomwordssample/MainActivity.java
package dev.emg.roomwordssample;

import android.content.Intent;
import android.os.Bundle;
import android.widget.Toast;
import androidx.annotation.Nullable;
import androidx.lifecycle.Observer;
import androidx.lifecycle.ViewModelProvider;
import androidx.lifecycle.ViewModelProviders;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import com.google.android.material.floatingactionbutton.FloatingActionButton;
import com.google.android.material.snackbar.Snackbar;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import android.view.View;
import android.view.Menu;
import android.view.MenuItem;
import dev.emg.roomwordssample.adapters.WordListAdapter;
import dev.emg.roomwordssample.model.Word;
import dev.emg.roomwordssample.viewmodel.WordViewModel;
import java.util.List;
import java.util.Objects;

public class MainActivity extends AppCompatActivity {

  private WordViewModel mWordViewModel;
  public static final int NEW_WORD_ACTIVITY_REQUEST_CODE = 1;

  @Override
  protected void onCreate(Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    setContentView(R.layout.activity_main);
    Toolbar toolbar = findViewById(R.id.toolbar);
    setSupportActionBar(toolbar);

    RecyclerView rv = findViewById(R.id.recyclerView);
    final WordListAdapter adapter = new WordListAdapter(this);
    rv.setAdapter(adapter);
    rv.setLayoutManager(new LinearLayoutManager(this));

    mWordViewModel = ViewModelProviders.of(this).get(WordViewModel.class);
    mWordViewModel.getAllWords().observe(this, new Observer<List<Word>>() {
      @Override public void onChanged(List<Word> words) {
        adapter.setWords(words);
      }
    });

    FloatingActionButton fab = findViewById(R.id.fab);
    fab.setOnClickListener(new View.OnClickListener() {
      @Override
      public void onClick(View view) {
        Intent intent = new Intent(MainActivity.this, NewWordActivity.class);
        startActivityForResult(intent, NEW_WORD_ACTIVITY_REQUEST_CODE);
      }
    });
  }

  @Override
  protected void onActivityResult(int requestCode, int resultCode, @Nullable Intent data) {
    super.onActivityResult(requestCode, resultCode, data);
    if (requestCode == NEW_WORD_ACTIVITY_REQUEST_CODE && resultCode == RESULT_OK) {
      assert data != null;
      Word word =
          new Word(Objects.requireNonNull(data.getStringExtra(NewWordActivity.EXTRA_REPLY)));
      mWordViewModel.insert(word);
    } else {
      Toast.makeText(getApplicationContext(), R.string.empty_not_saved, Toast.LENGTH_SHORT).show();
    }
  }

  @Override
  public boolean onCreateOptionsMenu(Menu menu) {
    // Inflate the menu; this adds items to the action bar if it is present.
    getMenuInflater().inflate(R.menu.menu_main, menu);
    return true;
  }

  @Override
  public boolean onOptionsItemSelected(MenuItem item) {
    // Handle action bar item clicks here. The action bar will
    // automatically handle clicks on the Home/Up button, so long
    // as you specify a parent activity in AndroidManifest.xml.
    int id = item.getItemId();

    //noinspection SimplifiableIfStatement
    if (id == R.id.action_settings) {
      return true;
    }

    return super.onOptionsItemSelected(item);
  }
}

File: Powerful-Android-Apps/app/src/main/java/dev/emg/powerfulandroid/ui/main/MainActivity.kt
package dev.emg.powerfulandroid.ui.main

import android.content.Intent
import android.os.Bundle
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.Observer
import dev.emg.powerfulandroid.BaseApplication
import dev.emg.powerfulandroid.databinding.ActivityMainBinding
import dev.emg.powerfulandroid.di.main.MainComponent
import dev.emg.powerfulandroid.session.SessionManager
import dev.emg.powerfulandroid.ui.auth.AuthActivity
import timber.log.Timber
import javax.inject.Inject

class MainActivity : AppCompatActivity() {

  lateinit var mainComponent: MainComponent
  @Inject
  lateinit var sessionManager: SessionManager

  private lateinit var binding: ActivityMainBinding

  override fun onCreate(savedInstanceState: Bundle?) {

    mainComponent = (application as BaseApplication).appComponent.mainComponent()
      .create()
    mainComponent.inject(this)

    super.onCreate(savedInstanceState)
    binding = ActivityMainBinding.inflate(layoutInflater)
    val view = binding.root
    setContentView(view)

    binding.toolBar.setOnClickListener {
      sessionManager.logout()
    }

    subscribeObservers()
  }

  fun subscribeObservers() {
    Timber.d("subscribeObservers()... is called")
    sessionManager.cachedToken.observe(this, Observer { authToken ->
      if (authToken == null || authToken.account_pk == -1 || authToken.token == null) {
        navAuthActivity()
        finish()
      }
    })
  }

  private fun navAuthActivity() {
    Timber.d("navAuthActivity()... is called")
    val intent = Intent(this, AuthActivity::class.java)
    startActivity(intent)
    finish()
  }

}
